picard/tagger.py

```diff
@@ -511,81 +544,117 @@ def add_files(self, filenames, target=None):
-                return (os.path.join(root, f) for f in files)
-
-        def process(result=None, error=None):
-            if result:
-                if error is None:
-                    self.add_files(result)
-                thread.run_task(get_files, process)
-
-        process(True, False)
-
-    def _add_directory_non_recursive(self, path):
+                time.sleep(0.01)
+
+    def _scan_dir(self, ignore_hidden, folders, recursive):
         files = []
-        for f in os.listdir(path):
-            listing = os.path.join(path, f)
-            if os.path.isfile(listing):
-                files.append(listing)
-        number_of_files = len(files)
-        if number_of_files:
-            mparms = {
-                'count': number_of_files,
-                'directory': path,
-            }
-            log.debug("Adding %(count)d files from '%(directory)r'" %
-                      mparms)
-            self.window.set_statusbar_message(
-                ngettext(
-                    "Adding %(count)d file from '%(directory)s' ...",
-                    "Adding %(count)d files from '%(directory)s' ...",
-                    number_of_files),
-                mparms,
-                translate=None,
-                echo=None
-            )
-            # Function call only if files exist
+        local_folders = list(folders)
+        while local_folders:
+            current_folder = local_folders.pop(0)
+            current_folder_files = []
+            for entry in os.scandir(current_folder):
+                if ignore_hidden and is_hidden(entry.path):
+                    continue
+                if recursive and entry.is_dir():
+                    local_folders.extend([entry.path])
+                else:
+                    current_folder_files.extend([entry.path])
+
+            number_of_files = len(current_folder_files)
+            if number_of_files:
+                mparms = {
+                    'count': number_of_files,
+                    'directory': current_folder,
+                }
+                log.debug("Adding %(count)d files from '%(directory)r'" %
+                          mparms)
+                self.window.set_statusbar_message(
+                    ngettext(
+                        "Adding %(count)d file from '%(directory)s' ...",
+                        "Adding %(count)d files from '%(directory)s' ...",
+                        number_of_files),
+                    mparms,
+                    translate=None,
+                    echo=None
+                )
+                files.extend(current_folder_files)
+        return files
+
+    def add_directory(self, path):
+        Thread(target=self._add_directory, args=[path, config.setting['recursively_add_files']]).start()
+
+    def _add_directory(self, path, recursive=False):
+        if not os.path.exists(path) or not os.path.isdir(path):
+            return
+
+        ignore_hidden = config.setting["ignore_hidden_files"]
```

Comment:
Perhaps we should pass this as parameter too (as `recursive`).

Reply:
Done.